### PR TITLE
MA-2434 : Migrate IAM Role Creation logic From Sails API(AwsAccountController.js) into StepFunction

### PR DESCRIPTION
**Pull request by chandramishra**
Purpose: Migrate IAM Role Creation logic From Sails API(AwsAccountController.js) into StepFunction

For more details Please refer this link : https://mountdiablo.jira.com/browse/MA-2434

Changes: API & AWS GIT HUB MicroServices

‌

Unit Testing

NA

ES Lint:

Implemented

### CODE REVIEW

File: src/index_create_aws_iam_roles.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  const dataDogPolicyDoc = JSON.parse(
+    fs.readFileSync(`${__dirname}/json/datadog-integration_policy.json`, {
+      encoding: 'utf8'
+    })
+  );
+
+  options.account = event.final_result.account_id;
+  options.assumeRolePolicyDocument = awsroles.assumeRolePolicyDocument;
+  options.assumeRolePolicyDocument.Statement[0].Principal.AWS =
+    'arn:aws:iam::442294194136:role/federate';
```
Comment (RajkumarShivage) [inline src/index_create_aws_iam_roles.js]: is it hardcoded?

File: src/lib/awsIamRole.js [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+            }
+            cb(null, options);
+        });
+    });
+}
+*/
+var createRole = function(iam, options, cb) {
+  var extid = options.externalId;
+  options.assumeRolePolicyDocument.Statement[0].Condition.StringEquals = {"sts:ExternalId": extid};
+  //options=JSON.parse(JSON.stringify(options));
```
Comment (RajkumarShivage) [inline src/lib/awsIamRole.js]: if not required remove these lines.

File: src/index_create_aws_iam_roles.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      if (payload.federate) {
+        dbIamRoles.push({
+          account: payload.account,
+          externalId: payload.externalId,
+          path: payload.path,
+          roleName: payload.roleName
+        });
+      }
+      awsIamRole.createRole(payload, (err, data) => {
+        console.log(err);
```
Comment (RajkumarShivage) [inline src/index_create_aws_iam_roles.js]: is if condition missing here?

File: src/index_create_aws_iam_roles.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      let payload = {};
+      Object.assign(payload, options, roles[i]);
+      payload.externalId = uuid.v4();
+      payload = JSON.parse(JSON.stringify(payload));
+      if (payload.roleName == 'DatadogAWSIntegrationRole') {
+        payload.PolicyDocument = dataDogPolicyDoc;
+        payload.assumeRolePolicyDocument.Statement[0].Principal.AWS = `arn:aws:iam::${event
+          .account.billingDetails.datadogAwsId}:root`;
+      }
+      if (payload.federate) {
```
Comment (SonalOjha) [inline src/index_create_aws_iam_roles.js]: What if the federate attribute is not present?

File: src/lib/awsIamRole.js [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+var findRolePolicy = function(iam, options, cb) {
+  var params = {
+    RoleName: options.roleName
+  };
+  iam.listRolePolicies(params, function(err, data) {
+    if (err) {
+      return cb("Failed to find a policy of the newly created role : " + err);
+    }
+    else {
+      if (data.PolicyNames.lenth > 0 && data.PolicyNames[0] == options.policyArn) {
```
Comment (SonalOjha) [inline src/lib/awsIamRole.js]: it should be length here

File: src/lib/awsIamRole.js [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+        if(options.onboardAccount){
+          cb(null, options.roleArn);
+        } else{
+          getUser(iam, options, cb);
+        }
+
+      }
+      else {
+        attachRolePolicy(iam, options, cb);
```
Comment (SonalOjha) [inline src/lib/awsIamRole.js]: Is it correct to attach role policy while listing ?

**Comment by RajkumarShivage** [comment]
Seperate IAM role creation functionality in two parts 1. Federation IAM roles 2. Integration IAM roles

**Review by bhargavapratik — APPROVED** [review]
LGTM

File: src/json/federate-role_info.json [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+				"roleName": "FullAdmin",
+				"policyArn": "arn:aws:iam::aws:policy/AdministratorAccess",
+				"federate": true
+			},
+			{
+				"roleName": "DatadogAWSIntegrationRole",
+				"policyArn": "",
+				"federate": false
+			}, {
+				"roleName": "sgas-dev-admin",
```
Comment (bhargavapratik) [inline src/json/federate-role_info.json]: sgas-dev-admin role name should not be hard coded here.

Comment (chandramishra) [reply src/json/federate-role_info.json]: address this part in MA-2435

**Review by RajkumarShivage — APPROVED** [review]
LGTM

**Review by SonalOjha — APPROVED** [review]

**Pull request merged**